Source/DafnyStandardLibraries/src/DafnyStdLibs/DynamicArray.dfy

```diff
@@ -0,0 +1,244 @@
+module DafnyStdLibs.DynamicArray {
+  import opened BoundedInts
+  import opened Wrappers
+
+  /**
+  The `DynamicArray` module and class define a data structure that has the same performance characteristics of an array, 
+  but additionally allows growing and shrinking the size of the array. The cost of shrinking the array is constant time, 
+  while the cost of growing it by a single element is either constant time or linear in the size of the array. 
+  When growing the array multiple times, the amortized cost of growing the array by one element is constant time.
+
+  The downsize of using a DynamicArray is that it occupies up to twice the amount of memory as its size, 
+  and when elements are removed, its occupied memory does not decrease. 
+  */
+  class DynamicArray<A> {
+    ghost var items : seq<A>
+    ghost var Repr : set<object>
+
+    const defaultValue: A
+    var size: uint32
+    var capacity: uint32
+    var data: array<A>
+
+    const MAX_CAPACITY: uint32 := (TWO_TO_THE_32 - 1) as uint32
+    const MAX_CAPACITY_BEFORE_DOUBLING: uint32 := MAX_CAPACITY / 2
+
+    ghost predicate Valid?()
+      reads this, Repr
+      ensures Valid?() ==> this in Repr
+    {
+      && Repr == {this, data}
+      && capacity != 0
+      && data.Length == capacity as int
+      && size <= capacity
+      && size as int == |items|
+      && items == data[..size]
+    }
+
+    constructor(defaultValue: A, initial_capacity: uint32 := 8)
```

Comment:
> You can avoid this by using a ghost datatype constructor, see TEST/comp/Uninitialized.dfy

I don't understand how this trick works. Can I initialize a non-ghost array variable with ghost values? I tried `new MaybeInitialized<A>[initialCapacity]` but I get the message `unless an initializer is provided for the array elements, a new array of 'MaybeInitialized<A>' must have empty size`

If I use `new MaybeInitialized<A>[initialCapacity](_ => Uninitialized)`, then I get `ghost constructor is allowed only in specification contexts`

The test you linked worked because it does not use `--enforce-determinism`, while this file does use that.

Reply:
The trick uses the first of these, that is:

``` dafny
new MaybeInitialized<A>[initialCapacity]
```

without any further initializer. This means that the compiler is free to initialize the new array elements any way it wants, as long as the values it uses are of the right type. In fact, the compiler is also allowed to pick the `Uninitialized` value, which you are not allowed to mention yourself in non-ghost code.

However, if the `new` allocation does not give an initializer, then the statement is nondeterministic.

Why does this example use `--enforce-determinism`?